redis/src/cluster_async/connections_logic.rs

```diff
@@ -0,0 +1,465 @@
+    cluster::get_connection_info,
+    cluster_client::ClusterParams,
+    cluster_topology::MANAGEMENT_CONN_NAME,
+    ErrorKind, IntoConnectionInfo, RedisError, RedisFuture, RedisResult,
+};
+
+use futures::prelude::*;
+use futures_time::future::FutureExt;
+
+use tracing::warn;
+
+use super::{AsyncClusterNode, Connect};
+
+#[doc(hidden)]
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum RefreshConnectionType {
+    // Refresh only user connections
+    OnlyUserConnection,
+    // Refresh only management connections
+    OnlyManagementConnection,
+    // Refresh all connections: both management and user connections.
+    AllConnections,
+}
+
+impl Connect for MultiplexedConnection {
+    fn connect<'a, T>(
+        info: T,
+        socket_addr: Option<SocketAddr>,
+    ) -> RedisFuture<'a, (MultiplexedConnection, Option<IpAddr>)>
+    where
+        T: IntoConnectionInfo + Send + 'a,
+    {
+        async move {
+            let connection_info = info.into_connection_info()?;
+            let client = crate::Client::open(connection_info)?;
+
+            #[cfg(feature = "tokio-comp")]
+            return client
+                .get_multiplexed_async_connection_inner::<crate::aio::tokio::Tokio>(socket_addr)
+                .await;
+
+            #[cfg(all(not(feature = "tokio-comp"), feature = "async-std-comp"))]
+            return client
+                .get_multiplexed_async_connection_inner::<crate::aio::async_std::AsyncStd>(
+                    socket_addr,
+                )
+                .await;
+        }
+        .boxed()
+    }
+}
+
+/// Return true if a DNS change is detected, otherwise return false.
+/// This function takes a node's address, examines if its host has encountered a DNS change, where the node's endpoint now leads to a different IP address.
+/// If no socket addresses are discovered for the node's host address, or if it's a non-DNS address, it returns false.
+/// In case the node's host address resolves to socket addresses and none of them match the current connection's IP,
+/// a DNS change is detected, so the current connection isn't valid anymore and a new connection should be made.
+pub(crate) async fn has_dns_changed(addr: &str, curr_ip: &IpAddr) -> bool {
+    let (host, port) = match get_host_and_port_from_addr(addr) {
+        Some((host, port)) => (host, port),
+        None => return false,
+    };
+    let mut updated_addresses = match get_socket_addrs(host, port).await {
+        Ok(socket_addrs) => socket_addrs,
+        Err(_) => return false,
+    };
+
+    !updated_addresses.any(|socket_addr| socket_addr.ip() == *curr_ip)
+}
+
+fn warn_mismatch_ip(addr: &str, new_ip: Option<IpAddr>, prev_ip: Option<IpAddr>) {
+    warn!(
+        "New IP was found for node {:?}: 
+                new connection IP = {:?}, previous connection IP = {:?}",
+        addr, new_ip, prev_ip
+    );
+}
+
+fn warn_management_conn_faild(addr: &str, err: RedisError) {
```

Comment:
Because if the user connection failed we return an error. When only the management connection failed we return the node without it, so we need to have some warning indicating it.